tests/mir-opt/sroa/structs.rs

```diff
@@ -76,31 +128,88 @@ fn f(a: *const u32) {
 // of them to `f`. However, this would lead to a miscompilation because `b` and `c`
 // might no longer appear right after `a` in memory.
 pub fn escaping() {
+    // CHECK-LABEL: fn escaping(
+
+    // CHECK-DAG: [[struct:_[0-9]+]]: Escaping;
+    // CHECK-DAG: [[ptr:_[0-9]+]]: *const u32;
+    // CHECK-DAG: [[ref:_[0-9]+]]: &u32;
+    // CHECK-DAG: [[a:_[0-9]+]]: u32;
+
+    // CHECK: bb0: {
+    // CHECK: [[struct]] = Escaping {
+    // CHECK: [[ref]] = &([[struct]].0
+    // CHECK: [[ptr]] = &raw const (*[[ref]]);
     f(&Escaping { a: 1, b: 2, c: g() }.a);
 }
 
+/// Check that copies from an internal struct are destructured and reassigned to
+/// the original struct.
 fn copies(x: Foo) {
+    // CHECK-LABEL: fn copies(
+
+    // CHECK: [[external:_[0-9]+]]: Foo) ->
+    // CHECK: [[internal:_[0-9]+]]: Foo;
+    // CHECK-DAG: [[byte:_[0-9]+]]: u8;
+    // CHECK-DAG: [[unit:_[0-9]+]]: ();
+    // CHECK-DAG: [[str:_[0-9]+]]: &str;
+    // CHECK-DAG: [[opt_isize:_[0-9]+]]: std::option::Option<isize>;
+
+    // CHECK: bb0: {
+    // CHECK-DAG: [[byte]] = ([[external]].0
+    // CHECK-DAG: [[unit]] = ([[external]].1
+    // CHECK-DAG: [[str]] = ([[external]].2
+    // CHECK-DAG: [[opt_isize]] = ([[external]].3
+
     let y = x;
     let t = y.a;
     let u = y.c;
     let z = y;
     let a = z.b;
 }
 
+/// Check that copies from an internal struct are destructured and reassigned to
+/// the original struct.
 fn ref_copies(x: &Foo) {
+    // CHECK-LABEL: fn ref_copies(
+
+    // CHECK: [[external:_[0-9]+]]: &Foo) ->
+    // CHECK: [[internal:_[0-9]+]]: Foo;
+    // CHECK-DAG: [[byte:_[0-9]+]]: u8;
+    // CHECK-DAG: [[unit:_[0-9]+]]: ();
+    // CHECK-DAG: [[str:_[0-9]+]]: &str;
+    // CHECK-DAG: [[opt_isize:_[0-9]+]]: std::option::Option<isize>;
+
+    // CHECK: bb0: {
+    // CHECK-DAG: [[byte]] = ((*[[external]]).0
+    // CHECK-DAG: [[unit]] = ((*[[external]]).1
+    // CHECK-DAG: [[str]] = ((*[[external]]).2
+    // CHECK-DAG: [[opt_isize]] = ((*[[external]]).3
+
     let y = *x;
     let t = y.a;
     let u = y.c;
 }
 
+/// Check that deaggregated assignments from constants are placed after the constant's
+/// assignment. Also check that copying field accesses from the copy of the constant are
+/// reassigned to copy from the constant.
 fn constant() {
+    // CHECK-LABEL: constant(
+
+    // CHECK: [[constant:_[0-9]+]]: (usize, u8);
+
+    // CHECK: bb0: {
+    // CHECK-NOT: scope
+    // CHECK-NOT: [[constant]]
+    // CHECK: [[constant]] = const
     const U: (usize, u8) = (5, 9);
     let y = U;
     let t = y.0;
     let u = y.1;
```

Comment:
Could you add the patterns for the assignments from `[[constant]]` to those locals?

Reply:
done